README.md
# RAINBOW-Distributed data storage

Rainbow's distributed data storage is developed using the [**Apache Ignite**](https://ignite.apache.org/) main-memory database.

The storage component comprises of 2 instances, the simple server node and the cluster head.

## Server

The Server is responsible for storing local data and/or remote data depending on other nodes' condition. 

It incorporates the services:

- Http Service: which is the API for ingesting and extracting data for RAINBOW services.
- Data movement service: which is responsible for transferring ingestion and extraction requests of monitoring data. It is also responsible for replicating local data to remote nodes if the cluster head requests it.
- Data management service: which is a low-level service responsible for directly writing/reading data to the different local caches.

### Http service

The Http Service is responsible for handling requests from external services and provides the following endpoints:

- `/nodes`: `POST` request that returns the list of active storage instances with their hostnames and their type of instance.
- `/put`: `POST` request for ingestion of monitoring data.
- `/get`: `POST` request that returns monitoring data with their values.
- `/query`: `POST` request that returns an aggregated value from the monitoring data. 
- `/list`: `POST` request that returns a list of the monitoring metadata. 
- `/monitoring`: `DELETE` request that deletes the specified monitoring data.
- `/analytics/put`: `POST` request for ingestion of analytics data.
- `/analytics/get`: `POST` request that returns analytics data with their values.
- `/analytics`: `DELETE` request that deletes the specified analytics data.
- `/app/put`: `POST` request for ingestion of app data.
- `/app/get`: `POST` request that returns app data with their values.
- `/app`: `DELETE` request that deletes the specified app data.

## Cluster head

The cluster head instance is responsible for monitoring the Ignite cluster and moving monitoring data when deemed necessary. It is also used to store and extract data as a simple server node. 

It incorporates all the services of the simple server node with the addition of 1 service: 

- Movement service: which is responsible for monitoring the restarts/fails of the Ignite cluster and deciding whether a node should start replicated its data points to a remote one. The process is automated and runs on specific time intervals. 

## Caches

Ignite uses caches for both persistent and in-memory data. Each server node along with the cluster head have 4 local caches for storage of both persistent and in-memory data.

- *LatestMonitoring* cache is used for in-memory storage of the latest values for every monitoring metric.
- *HistoricalMonitoring* cache is used for persistent storage of the historical values for every monitoring metric.
- *Analytics* cache is used for persistent storage of processed data from the Analytics component.
- *ApplicationData* cache is used for in-memory storage of simple key-value data with timestamps.

Along with the local caches, each node has a copy of the 3 replicated caches that are used from the placement and the movement service.

- *MetaMonitoring* cache is used for persistent storage of metadata for every monitoring metric and the entity they belong to.
- *Replicas* cache is used to store the remote nodes that a monitoring data point is replicated.
- *Restarts* cache is used to store the restarts/fails of the server nodes in the Ignite cluster.

### Data Regions

2 data regions are available with specified off-heap memory limits. 

- *Persistent_region*: is used for the persistent caches with default max off-heap memory size to 256MB.
- *Default_region*: is the region for every in-memory cache with default max off-heap memory size to 256MB.

The default total size of the regions is 512MB split evenly between them (256MB for persistent region and 256MB for in-memory). The user can optionally change the total memory size with the **SIZE** environment variable. *The minimum off-heap size for each region is 100MB.*

## Deployment

In an **actual deployment** a single cluster head instance needs to be deployed first in order to create the cluster and the necessary caches. If many instances are deployed concurrently some of them may not be able to enter the cluster since each of them tries to create one.

The docker image can be created using the `build_image.sh` script in the respective folder.

## Configuration

Environment variables control the optional features such as persistence and the instance type (server, client). A list of all available variables that can be used in the docker container is below:

1. **CLUSTER_HEAD**: The variable that controls the instance type. Available values are "**true**" and "**false**". *Default value is "**false**"*.
2. **HOSTNAME**: The variable that is used for the container's hostname. If it is skipped, the program tries to find its own hostname using the `InetAddress` library.
3. **DISCOVERY**: The variable that controls the discovery process. It should be a comma separated list of hostnames, e.g "**server-1,server-2**".
4. **APP_CACHE**: The variable that controls whether the user-application cache is on or off. Available values are "**true**" and "**false**". *Default value is "**false**"*.
5. **EVICTION**: The variable that controls the eviction period for in-memory data. The value represents the eviction rate in hours from the creation of a data row. *Default value is "**168** hours (1 week)"*.
6. **SIZE**: The variable that controls the total off-heap size for the data regions. Each region gets half of the total size as the maximum off-heap memory size.
7. **JAVA_OPTS**: The variable that controls different JVM options. The user can change the max JVM heap memory size (e.g. "-Xms256m -Xmx256m").
8. **PLACEMENT**: The variable that controls whether the placement service will be on or off. Available values are "**true**" and "**false**". *Default value is "**true**"*.

## Docker ports

The available ports that are exposed from the Docker deployment through the above process are:

- 50000: REST API for the Server instance.
- 47500: Used for the discovery of the cluster.
- 47100: Used for the communication of the instances.

## REST API examples

* For the monitoring data, a `/put` **POST request** needs to have a similar body:

```
{"monitoring": [
    {   
        "entityID": "ent1",
        "entityType": "fog",
        "metricID": "metr1",
        "name": "cpu",
        "units": "units1",
        "desc": "CPU",
        "group": "group1",
        "minVal": 5,
        "maxVal": 10,
        "higherIsBetter": true,
        "val": 6,
        "timestamp": 1611318068003,
        "pod":{
            "uuid": "123",
            "name": "kube",
            "namespace": "namespace"
        },
        "container":{
            "id": "345",
            "name": "container"
        }
    }
]}
```

The `entityID`, `metricID`, `val` and `timestamp` fields are mandatory.

* For the monitoring data, a `/get` **POST request** needs to have a similar body:

```
{   
    "metricID": ["metr1"], (optional)
    "entityID": ["ent1"], (optional)
    "entityType": ["fog"], (optional)
    "groupName": ["group1"], (optional)
    "podName": ["pod1"], (optional)
    "podNamespace": ["name1"], (optional)
    "containerName": ["container1"], (optional)
    "from": 1611318068000, (optional)
    "to": 2611318068000, (optional)
    "nodes": ["ip1","ip2"] (optional)
}
```
The `from` and `to` variables are optional. If they are missing only the latest value will be returned. The filters `metricID`, `entityID`, `podName`, `podNamespace` and `containerName` can be skipped or present in any combination for filtering out the results.

Using the optional `nodes` keyword with an array of ips (or empty array) returns the requested metric from a set of server nodes (or every active server node):

An example **response** body is:

```
{
    "monitoring": [
        {
            "node": "ip1",
            "data": [
                {
                    "metricID": "metr1",
                    "entityID": "ent1",
                    "values": [
                        {
                            "timestamp": 3,
                            "val": 30.0
                        }
                    ],
                    "entityType": "fog",
                    "name": "myname",
                    "units": "myunits",
                    "desc": "this is great",
                    "group": "mygroup",
                    "minVal": 5.0,
                    "maxVal": 10.0,
                    "higherIsBetter": true,
                    "pod": {
                        "uuid": "123",
                        "name": "kube",
                        "namespace": "namespace"
                    },
                    "container": {
                        "id": "345",
                        "name": "container"
                    }
                },
                {...},
                {...}
            ]
        },
        {...}
    ]
}
```

* For the aggregated query, a `\query` **POST request** needs to have the same body with the `\get` request in addition to one of the aggregated functions:

```
"agg": "sum"/"max"/"min"/"avg"

```

An example **response** body is:

```
{
    "value": 30.0
}
```

* For the list of monitoring data, a `/list` **POST request** needs to have a similar body:

```
{
    "metricID": ["metr1"], (optional)
    "entityID": ["ent1"], (optional)
    "podName": ["pod1"], (optional)
    "podNamespace": ["name1"], (optional)
    "containerName": ["container1"], (optional)
    "nodes": ["ip1"], (optional)
    "fields": ["metricfda", "entityType", "containerName", "podName", "desc", "podUUID", "higherIsBetter"] (optional)
}
```

The filters and the node list are used in the same way as the `\get` request for monitoring data. The *fields* keyword is used to reduce the metadata that are returned.

An example **response** body is:

```
{
    "metric": [
        {
            "node": "ip1",
            "metricID": "metr1",
            "entityID": "as1",
            "entityType": "fog",
            "higherIsBetter": true,
            "desc": "this is great",
            "pod": {
                "podName": "kube",
                "podUUID": "123"
            },
            "container": {
                "containerName": "container"
            }
        }
    ]
}
```

* For the analytics data, a `/analytics/put` **POST request** needs to have a similar body:

```
{"analytics": [
    {   
        "key": "app1",
        "val": "app1data",
        "timestamp": 1611318068000
    }
]}
```

All 3 fields are mandatory.

* For the analytics data, a `/analytics/get` **POST request** needs to have a similar body:

```
{   
    "key": ["key1"], (optional)
    "from": 1611318068000, (optional)
    "to": 2611318068000 (optional)
    
}
```

An example **response** body is:

```
{
    "analytics": [
        {
            "key": "key1",
            "values": [
                {
                    "timestamp": 1611318068000,
                    "val": 3.0
                }
            ]
        }
    ]
}
```

* The application-specific endpoints work the **same as the analytics ones**.

CHANGELOG.md
# Changelog

## [1.4.10] - 23-02-2022

- New filter on "group" field of monitoring metrics

## [1.4.9] - 07-12-2022

- Fixed timestamp precision on data placement algorithm
- New unstable nodes will be removed as replication destinations from previous iterations of the placement algorithm
- New stable nodes will be removed as replication sources from previous iterations of the placement algorithm

## [1.4.8] - 01-03-2022

- Added order by timestamp on data extraction
- Fixed response time for analytics

## [1.4.7] - 30-03-2022

- Hotfix on json schema

## [1.4.6] - 23-03-2022

### Added

- New filter on monitoring requests
- Option to get monitoring data from every node with an empty `node` field

## [1.4.5] - 18-03-2022

### Changed

- Cost function on unstable nodes
- Unstable nodes can a max capacity on outgoing replicas

## [1.4.4] - 28-02-2022

### Changed

- Fixed bug with ips/hostnames
- Fixes based on sonarqube
- Fix on instance name

## [1.4.0] - 21-02-2022

### Added

- New endpoints for deleting data (monitoring, analytics, app).
- New endpoint for getting the list of monitoring metadata.
- New endpoint for getting list of active nodes
- Aggregation queries (and endpoint) on monitoring data.
- New service in charge of data movement for monitoring data. I.e. getting data from remote nodes and replicating local data to remote ones. 
- New service in charge of data placement based on node restarts/fails (can be optionally turned off using an environment variable).
- Two new replicated caches that help store metadata for the placement and movement services.
- New environment variable `CLUSTER_HEAD` which creates the main server node that initialize the caches and runs the placement service.

### Changed

- Monitoring metadata cache is now replicated to all nodes.
- Changed analytics cache to be able to store historical data.
- Split http service from the data management and movement services.
- Updated unit tests
- Refactoring of the codebase with SonarLint recommendations.

### Removed

- Client option is removed since the cluster head option does the same job and stores analytics data.

## [1.3.5] - 23-11-2021

### Added

- Wildcards on filters for get queries.

## [1.3.4] - 09-11-2021

### Added

- New fields in the metadata of the monitoring data about the pod and container info.
- Added the unique combination of `metricID` and `entityID` as key to every stored cache of monitoring data.
- New filters on `podName`, `podNamespace` and `containerName` based on the new metadata on the `get` request.

### Changed

- Changed the `metricID` and `entityID` filters on the `get` request for the monitoring data to be used alongside the new ones in any combination.

### Removed

- The `latest` field from the `get` request on the monitoring data. If no `from` and `to` fields are present then the latest data are returned.

## [1.3.3] - 18-10-2021

### Added

- JVM options on dockerfiles for GC and heap memory. Optional user input for heap memory throught the JAVA_OPTS environment variable.
- Added option for monitoring and analytics `get` commands to return data from a set (or all) server nodes based on their hostnames/ips.

### Changed

- Optional user input for regions total size. Default total size is 512MB which is split evenly between the in-memory and persistent regions.

## [1.3.2] - 11-10-2021

### Added

- Maximum size (500MB) for default data region (for latest and analytics data).
- Maximum size (1GB) for persistent data region (for historical and meta data).
- New region with maximum size of 200MB for application data.
- Gitlab CI/CD pipeline

### Changed

- Persistent is no longer optional. Historical and Metadata caches are persistent-only.
- Dockerfile for gitlab CI/CD pipeline

## [1.3.1] - 01-06-2021

### Added 

- In-memory cache for processed analytic results using a key-value model.
- REST API routes for ingesting and extracting analytics data.
- New optional environment variable for the container's hostname.
 
## [1.3.0] - 22-05-2021

### Added 

- Client REST API to extract monitoring data from all ignite server nodes.
- In-memory key-value cache for user application data.
- Persistence options for historical and metadata caches.
- Eviction options for historical and metadata caches.
- Unit tests for the main functions of the Server instances (ingest, extract monitoring data).

### Changed 

- Merged the docker images (client and server) into one image.
- The docker image starts with a base of JDK 1.8 and Alpine instead of Ignite.
- Ignite is initialized through the code main function.
- Server or Client nodes are decided through an environment variable.
- Configuration of instances is implemented inside the code instead of XML files.
- Changed the `DataService` interface and implementation to incorporate 2 public classes that can be used from both server and client APIs.

## [1.2.2] - 04-02-2021 (bugfix)

### Changed 

- Fixed bug on sql queries with WHERE clause.
- Changed zookeeper service name.

## [1.2.1] - 29-01-2021 (hotfix)

### Changed 

- Changed `get` filter to get all metrics when there is no metricID or entityID keyword.

## [1.2.0] - 29-01-2021

### Added 

- Caches for storing historical data and metadata for all monitoring metrics.
- New filters for the `get` endpoint of `Ignite-server` in order to filter out returned values based on metricID/entityID and time period or latest values.
- `docker-compose.yml` file to deploy the solution with a single Ignite service and a zookeeper.

## [1.1.0] - 22-01-2021

### Added 

- Class for metrics with metadata.

### Changed

- Ingestion and Extraction services of `Ignite-server` are incorporated into a single REST API instead of a simple socket connection.
- Ingestion service does not call rebalancing service on each ingest task.
- Stopped `Ignite-client`'s extraction service in order to incorporate it in a REST API.  

## [1.0.1] - 20-01-2021

### Added


src/main/java/org/auth/csd/datalab/DataService.java
package org.auth.csd.datalab;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.ignite.Ignite;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.cache.CacheMode;
import org.apache.ignite.cache.query.QueryCursor;
import org.apache.ignite.cache.query.ScanQuery;
import org.apache.ignite.cache.query.SqlFieldsQuery;
import org.apache.ignite.configuration.CacheConfiguration;
import org.apache.ignite.lang.IgniteBiPredicate;
import org.apache.ignite.resources.IgniteInstanceResource;
import org.apache.ignite.services.ServiceContext;
import org.auth.csd.datalab.common.helpers.*;
import org.auth.csd.datalab.common.interfaces.DataInterface;
import org.json.JSONArray;
import org.json.JSONObject;
import org.rapidoid.buffer.Buf;
import org.rapidoid.http.AbstractHttpServer;
import org.rapidoid.http.HttpStatus;
import org.rapidoid.http.HttpUtils;
import org.rapidoid.net.Server;
import org.rapidoid.net.abstracts.Channel;
import org.rapidoid.net.impl.RapidoidHelper;

import javax.cache.expiry.CreatedExpiryPolicy;
import javax.cache.expiry.Duration;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class DataService implements DataInterface {

    //TODO ENV variables
    private String latestCacheName = "LatestMonitoring";
    private String historicalCacheName = "HistoricalMonitoring";
    private String metaCacheName = "MetaMonitoring";
    private String appCacheName = "ApplicationData";
    private String persistenceRegion = "Persistent_Region";
    private int evictionHours = 168;
    @IgniteInstanceResource
    private Ignite ignite;
    /**
     * Reference to the cache.
     */
    private IgniteCache<String, TimedMetric> myLatest;
    private IgniteCache<MetricKey, Metric> myHistorical;
    private IgniteCache<MetaMetricKey, MetaMetric> myMeta;
    private IgniteCache<String, String> myApp = null;

    UUID localNode = null;
    private String delimiter = ".";
    private Server server;

    /**
     * {@inheritDoc}
     */
    public void init(ServiceContext ctx) {
        System.out.println("Initializing Data Service on node:" + ignite.cluster().localNode());
        //Get the cache that is designed in the config for the latest data
        myLatest = ignite.cache(latestCacheName);
        //Get the eviction time
        try {
            String evict = readEnvVariable("EVICTION");
            if (evict != null) evictionHours = Integer.parseInt(Objects.requireNonNull(readEnvVariable("EVICTION")));
        } catch (NumberFormatException e) {
            System.out.println("Input Eviction value cannot be parsed to Integer.");
        }
        //Create a cache for the meta data
        CacheConfiguration<MetaMetricKey, MetaMetric> metaCfg = new CacheConfiguration<>(metaCacheName);
        metaCfg.setIndexedTypes(MetaMetricKey.class, MetaMetric.class);
        metaCfg.setCacheMode(CacheMode.LOCAL);
        //Persistence and eviction rate
        metaCfg.setDataRegionName(persistenceRegion);
        metaCfg.setExpiryPolicyFactory(CreatedExpiryPolicy.factoryOf(new Duration(TimeUnit.HOURS, evictionHours)));
        metaCfg.setEagerTtl(true);
        myMeta = ignite.getOrCreateCache(metaCfg);
        //Create a cache for the historical data
        CacheConfiguration<MetricKey, Metric> historicalCfg = new CacheConfiguration<>(historicalCacheName);
        historicalCfg.setIndexedTypes(MetricKey.class, Metric.class);
        historicalCfg.setCacheMode(CacheMode.LOCAL);
        //Persistence and eviction rate
        historicalCfg.setDataRegionName(persistenceRegion);
        historicalCfg.setExpiryPolicyFactory(CreatedExpiryPolicy.factoryOf(new Duration(TimeUnit.HOURS, evictionHours)));
        historicalCfg.setEagerTtl(true);
        myHistorical = ignite.getOrCreateCache(historicalCfg);
        //Create optional cache for application data
        if (Objects.equals(readEnvVariable("APP_CACHE"), "true")) {
            CacheConfiguration<String, String> appCfg = new CacheConfiguration<>(appCacheName);
            metaCfg.setCacheMode(CacheMode.LOCAL);
            myApp = ignite.getOrCreateCache(appCfg);
        }
        localNode = ignite.cluster().localNode().id();
        server = new CustomHttpServer().listen(50000);
    }

    /**
     * {@inheritDoc}
     */
    public void execute(ServiceContext ctx) {
        System.out.println("Executing Data Service on node:" + ignite.cluster().localNode());
    }

    /**
     * {@inheritDoc}
     */
    public void cancel(ServiceContext ctx) {
        System.out.println("Stopping Data Service on node:" + ignite.cluster().localNode());
        server.shutdown();
    }

    //------------APP cache----------------
    private void ingestAppData(HashMap<String, String> data) {
        for (Map.Entry<String, String> entry : data.entrySet()) {
            myApp.put(entry.getKey(), entry.getValue());
        }
    }

    private HashMap<String, String> extractAppData(List<String> search) {
        HashMap<String, String> data = new HashMap<>();
        IgniteBiPredicate<String, String> filter;
        if (!search.isEmpty()) filter = (key, val) -> search.contains(key);
        else filter = null;
        myApp.query(new ScanQuery<>(filter)).forEach(entry -> {
            data.put(entry.getKey(),entry.getValue());
        });
        return data;
    }

    //------------MONITORING----------------
    private void ingestMetric(HashMap<String, InputJson> data) {
        for (Map.Entry<String, InputJson> entry : data.entrySet()) {
            myLatest.put(entry.getKey(), new TimedMetric(entry.getValue().val, entry.getValue().timestamp));
            MetricKey metricTmp = new MetricKey(entry.getValue());
            myHistorical.put(metricTmp, new Metric(entry.getValue().val));
            myMeta.put(new MetaMetricKey(entry.getValue()), new MetaMetric(entry.getValue()));
        }
    }

    @Override
    public HashMap<String, String> extractLatestData(List<String> search) {
        HashMap<String, String> data = new HashMap<>();
        IgniteBiPredicate<String, TimedMetric> filter;
        if (!search.isEmpty()) filter = (key, val) -> search.contains(key);
        else filter = null;
        myLatest.query(new ScanQuery<>(filter)).forEach(entry -> {
            data.put(entry.getKey(), "\"val\": " + entry.getValue().val + " , \"timestamp\": " + entry.getValue().timestamp);
        });
        return data;
    }

    @Override
    public String extractHistoricalData(String metric, Long min, Long max) {
        List<String> res = new ArrayList<>();
        SqlFieldsQuery sql = new SqlFieldsQuery("select metricID, timestamp, val from METRIC WHERE metricID = '" + metric + "' AND timestamp >= " + min + " AND timestamp <= " + max);
        try (QueryCursor<List<?>> cursor = myHistorical.query(sql)) {
            for (List<?> row : cursor) {
                Long time = (Long) row.get(1);
                Double value = (Double) row.get(2);
                res.add(" { \"timestamp\": " + time + " , \"val\": " + value + " } ");
            }
        }
        String result = "\"values\": [ " + String.join(", ", res) + " ] ";
        return result;
    }

    private HashMap<String, String> extractMeta(List<String> search) {
        HashMap<String, String> meta = new HashMap<>();
        for (String metric : search) {
            SqlFieldsQuery sql = new SqlFieldsQuery("select entityID, entityType, name, units, desc, groupName, minVal, maxVal, higherIsBetter from METAMETRIC WHERE metricID = '" + metric + "'");
            try (QueryCursor<List<?>> cursor = myMeta.query(sql)) {
                for (List<?> row : cursor) {
                    String res = " \"entityID\": \"" + row.get(0) + "\"" +
                            ", \"entityType\": \"" + row.get(1) + "\"" +
                            ", \"name\": \"" + row.get(2) + "\"" +
                            ", \"units\": \"" + row.get(3) + "\"" +
                            ", \"desc\": \"" + row.get(4) + "\"" +
                            ", \"group\": \"" + row.get(5) + "\"" +
                            ", \"minVal\": " + row.get(6) +
                            ", \"maxVal\": " + row.get(7) +
                            ", \"higherIsBetter\": " + row.get(8) + " ";
                    meta.put(metric, res);
                }
            }
        }
        return meta;
    }

    private ArrayList<String> getMetricID() {
        ArrayList<String> metrics = new ArrayList<>();
        SqlFieldsQuery sql = new SqlFieldsQuery("select metricID from METAMETRIC");
        try (QueryCursor<List<?>> cursor = myMeta.query(sql)) {
            for (List<?> row : cursor)
                metrics.add(row.get(0).toString());
        }
        return metrics;
    }

    private List<String> getMetricID(List<String> entities) {
        List<String> metrics = new ArrayList<>();
        for (String entity : entities) {
            SqlFieldsQuery sql = new SqlFieldsQuery("select metricID from METAMETRIC WHERE entityID = '" + entity + "'");
            try (QueryCursor<List<?>> cursor = myMeta.query(sql)) {
                for (List<?> row : cursor)
                    metrics.add(row.get(0).toString());
            }
        }
        return metrics;
    }

    //------------API----------------

    private class CustomHttpServer extends AbstractHttpServer {

        private final byte[] URI_PUT = "/put".getBytes();
        private final byte[] URI_GET = "/get".getBytes();
        //App cache
        private final byte[] URI_APP_PUT = "/app/put".getBytes();
        private final byte[] URI_APP_GET = "/app/get".getBytes();

        @Override
        protected HttpStatus handle(Channel ctx, Buf buf, RapidoidHelper req) {
            if (!req.isGet.value) {
                if (matches(buf, req.path, URI_PUT)) {
                    //Data structure for metrics
                    HashMap<String, InputJson> metrics = new HashMap<>();
                    //Read and parse json
                    try {
                        String body = buf.get(req.body);
                        JSONObject obj = new JSONObject(body);
                        //Get the monitoring keyword and parse the data
                        JSONArray monitor = obj.getJSONArray("monitoring");
                        for (int i = 0; i < monitor.length(); i++) {
                            JSONObject o = monitor.getJSONObject(i);
                            ObjectMapper m = new ObjectMapper();
                            try {
                                InputJson myMetric = m.readValue(o.toString(), InputJson.class);
                                metrics.put(myMetric.metricID, myMetric);
                            } catch (IOException e) {
                                e.printStackTrace();
                                return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("ERROR", "Error on json schema!"));
                            }
                        }
                        ingestMetric(metrics);
                        return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("OK", "Ingestion successful!"));
                    } catch (Exception e) {
                        e.printStackTrace();
                        return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("ERROR", "Error on data ingestion!"));
                    }
                }
                else if (matches(buf, req.path, URI_GET)) {
                    HashMap<String, String> result = new HashMap<>();
                    HashMap<String, String> meta = new HashMap<>();
                    //Read and parse json
                    String body = buf.get(req.body);
                    if (!body.equals("")) { //Check if body has filters
                        JSONObject obj = new JSONObject(body);
                        try {
                            ArrayList<String> ids = new ArrayList<>();
                            if (obj.has("entityID")) {
                                JSONArray entities = obj.getJSONArray("entityID");
                                List<String> entitiesList = entities.toList().stream().map(Object::toString).collect(Collectors.toList());
                                List<String> metrics = getMetricID(entitiesList);
                                ids.addAll(metrics);
                            } else if (obj.has("metricID")) {
                                JSONArray entities = obj.getJSONArray("metricID");
                                for (Object ent : entities) { //Store metric ids to list
                                    ids.add(ent.toString());
                                }
                            } else {
                                ids = getMetricID();
                            }
                            if (!ids.isEmpty()) {
                                if (obj.has("latest") && obj.getBoolean("latest")) { //Get only the latest data
                                    result = extractLatestData(ids);
                                } else if (obj.has("from") && obj.has("to")) {
                                    long from = obj.getLong("from");
                                    long to = obj.getLong("to");
                                    for (String metric : ids) {
                                        result.put(metric, extractHistoricalData(metric, from, to));
                                    }
                                }
                                //Get metadata
                                meta = extractMeta(ids);
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                            return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("ERROR", "Error on data extraction!"));
                        }
                    }
                    //Join meta with values
                    List<String> metaValues = new ArrayList<>();
                    for (String metric : result.keySet()) {
                        metaValues.add(" { \"metricID\": \"" + metric + "\" , " + result.get(metric) + " , " + meta.getOrDefault(metric, "") + " } ");
                    }
                    String finalRes = "{\"monitoring\": [ " + String.join(", ", metaValues) + " ]} ";
                    return json(ctx, req.isKeepAlive.value, finalRes.getBytes());
                }
                else if (myApp != null && matches(buf, req.path, URI_APP_PUT)){
                    //Data structure for app data
                    HashMap<String, String> data = new HashMap<>();
                    //Read and parse json
                    try {
                        String body = buf.get(req.body);
                        JSONObject obj = new JSONObject(body);
                        //Get the application keyword and parse the data
                        JSONArray application = obj.getJSONArray("application");
                        for (int i = 0; i < application.length(); i++) {
                            JSONObject o = application.getJSONObject(i);
                            if(o.has("key") && o.has("value")){
                                data.put(o.getString("key"), o.getString("value"));
                            }
                        }
                        ingestAppData(data);
                        return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("OK", "Ingestion successful!"));
                    } catch (Exception e) {
                        e.printStackTrace();
                        return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("ERROR", "Error on data ingestion!"));
                    }
                }
                else if (myApp != null && matches(buf, req.path, URI_APP_GET)) {
                    HashMap<String, String> result = new HashMap<>();
                    //Read and parse json
                    String body = buf.get(req.body);
                    if (!body.equals("")) { //Check if body has filters
                        JSONObject obj = new JSONObject(body);
                        try {
                            ArrayList<String> ids = new ArrayList<>();
                            if (obj.has("key")) {
                                JSONArray tmpKeys = obj.getJSONArray("key");
                                ids.addAll(tmpKeys.toList().stream().map(Object::toString).collect(Collectors.toList()));
                            }
                            result = extractAppData(ids);
                        } catch (Exception e) {
                            e.printStackTrace();
                            return serializeToJson(HttpUtils.noReq(), ctx, req.isKeepAlive.value, new Message("ERROR", "Error on data extraction!"));
                        }
                    }
                    //Join meta with values
                    List<String> metaValues = new ArrayList<>();
                    for (String metric : result.keySet()) {
                        metaValues.add(" { \"key\": \"" + metric + "\" , \"value\": \"" + result.get(metric) + "\" } ");
                    }
                    String finalRes = "{\"application\": [ " + String.join(", ", metaValues) + " ]} ";
                    return json(ctx, req.isKeepAlive.value, finalRes.getBytes());
                }
            }
            return HttpStatus.NOT_FOUND;
        }
    }

    /**
     * Method to read environment variables
     *
     * @param key The key of the variable
     * @return The variable
     */
    private String readEnvVariable(String key) {
        if (System.getenv().containsKey(key)) {
            String envVariable = System.getenv(key);
            return envVariable;
        } else return null;
    }

}
